onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2.cpp

```diff
@@ -0,0 +1,1059 @@
+/*++
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Licensed under the MIT License.
+
+Module Name:
+
+    sqnbitgemm_kernel_avx2.cpp.h
+
+Abstract:
+
+    This module implements the float/quantized n-bit integer matrix
+    multiplication kernels for x64 avx2.
+
+--*/
+
+#include <algorithm>
+#include <cassert>
+#include <utility>
+
+#include "sqnbitgemm.h"
+
+#include "sqnbitgemm_kernel_avx_common.h"
+#include "sqnbitgemm_kernel_avx_common_int8.h"
+
+
+MLAS_FORCEINLINE
+__m256
+load_float_n_avx2(const float* data, int n)
+{
+    assert(n <= 8);
+    if (n <= 0) {
+        return _mm256_setzero_ps();
+    }
+    static const int32_t mask_buffer[16] = {-1, -1, -1, -1, -1, -1, -1, -1, 0, 0, 0, 0, 0, 0, 0, 0};
+    const __m256i load_mask = _mm256_loadu_si256((const __m256i*)(mask_buffer + 8 - n));
+    return _mm256_maskload_ps(data, load_mask);
+}
+
+MLAS_FORCEINLINE void
+Q4BitBlkDequantBForSgemmBlkLen16_CompFp32_avx2(
+    float* FpData,
+    const std::byte* QuantBData,
+    const float* QuantBScale,
+    const std::byte* QuantBZeroPoint,
+    const size_t CountN,
+    const size_t CountK,
+    const size_t BlockCountK
+)
+{
+    constexpr size_t BlkLen16 = 16;
+    constexpr size_t BlkBitWidth4 = 4;
+
+    constexpr size_t blk_data_size_in_bytes = MlasQNBitBlkDataSizeInBytes(BlkBitWidth4, BlkLen16);
+    const size_t b_data_col_stride_in_bytes = BlockCountK * blk_data_size_in_bytes;
+    // TODO: constexpr use temaplte parameter
+    /*constexpr*/ const bool HasZeroPoint = QuantBZeroPoint != nullptr;
+    const size_t zp_col_stride_in_bytes = MlasQNBitZeroPointsForBlksSizeInBytes<BlkBitWidth4>(BlockCountK);
+
+    constexpr size_t NCols8 = 8;                   // process NCols8 columns of QuantB at a time
```

Comment:
>   constexpr size_t NCols8 = 8;  [](http://example.com/codeflow?start=1&length=32)

why choose 8 instead of 16? if use 16, it aligns with the GemmFloatKernel

Reply:
We will choose 16 in avx512 implementation in the next. current implementation keeps 2 x NCols8 __mm256 variables for weights and scales. if we choose 16, it will be too many. Agree we shall explore different implementation to see if we can process 16 cols at a time - 16 cols by 32 rows (load in with one instruction) of weights may be too many. For avx512, things will be different.